<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Data Preparation • nQuack</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Data Preparation">
<meta name="description" content="Learn how to prepare data for nQuack.
">
<meta property="og:description" content="Learn how to prepare data for nQuack.
">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-YWYQ5CX5NC"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YWYQ5CX5NC');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nQuack</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/BasicExample.html">Basic Example</a></li>
    <li><a class="dropdown-item" href="../articles/DataPreparation.html">Data Preparation</a></li>
    <li><a class="dropdown-item" href="../articles/ModelOptions.html">Model Options</a></li>
    <li><a class="dropdown-item" href="../articles/SimulateData.html">Simulate Data</a></li>
    <li><a class="dropdown-item" href="../articles/Outliers.html">Outliers</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Data Preparation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mgaynor1/nQuack/blob/HEAD/vignettes/DataPreparation.Rmd" class="external-link"><code>vignettes/DataPreparation.Rmd</code></a></small>
      <div class="d-none name"><code>DataPreparation.Rmd</code></div>
    </div>

    
    
<p>This vignette will teach you how to prepare raw sequence data for
input into nQuack. Our software requires sequences to be aligned to a
reference genome or target sequences.</p>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<div class="section level3">
<h3 id="align-to-a-reference-genome">Align to a reference genome<a class="anchor" aria-label="anchor" href="#align-to-a-reference-genome"></a>
</h3>
<p>After processing our raw sequence data, we have to align the reads to
a reference genome. Here we use <a href="https://github.com/bwa-mem2/bwa-mem2" class="external-link">bwa-mem2</a> to align our
reads. We then converted our SAM files to BAM files using <a href="http://www.htslib.org/" class="external-link">samtools</a>. This is our bash script:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Load module</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">module</span> load bwa-mem2/2.2.1</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="ex">module</span> load samtools/1.15</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co"># Index the reference</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="ex">bwa-mem2</span> index reference_genome.gz</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co"># Mapping </span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co">## -t Number of threads, here we run our mapping on 10 threads.  </span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co">## -M Indexed reference genome.</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="ex">bwa-mem2</span> mem <span class="at">-t</span> 10 <span class="at">-M</span> reference_genome.gz sample_001_1.fastq sample_001_2.fastq <span class="op">&gt;</span> sample_001.sam</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co"># SAM to BAM</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-S</span> <span class="at">-b</span> sample_001.sam <span class="op">&gt;</span> sample_001.bam</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="co"># SORT</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="ex">samtools</span> sort sample_001.bam <span class="at">-o</span> sample_001.bam</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="remove-repeats">Remove repeats<a class="anchor" aria-label="anchor" href="#remove-repeats"></a>
</h3>
<div class="section level4">
<h4 id="identify-repeat-regions">Identify repeat regions<a class="anchor" aria-label="anchor" href="#identify-repeat-regions"></a>
</h4>
<p>Prior to running samples through nQuack, we suggest repeat regions be
removed. To remove repeats, you first have to identify them. Here we
utilize <a href="https://www.pnas.org/doi/10.1073/pnas.1921046117" class="external-link">repeat
modeler</a> and <a href="https://www.repeatmasker.org/" class="external-link">repeat
masker</a> to identify and mask repeats. We also make a database of
these mask repeats, which will then be used to remove repeats from our
samples alignment.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co">## Load modules</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">module</span> load repeatmodeler/2.0</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="ex">module</span> load repeatmasker/4.1.1</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co"># Set database name</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="va">databasename</span><span class="op">=</span>Species</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co"># Build database based on the reference genome</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="ex">BuildDatabase</span> <span class="at">-name</span> <span class="va">$databasename</span> ReferenceGenome.fasta</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co"># Repeat Modeler</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">## -LTRStruct =  runs the LTR structural discovery pipeline ( LTR_Harvest and LTR_retreiver )</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co">### and combine results with the RepeatScout/RECON pipeline.</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="ex">RepeatModeler</span> <span class="at">-pa</span> 36 <span class="at">-database</span> <span class="va">$databasename</span> <span class="at">-LTRStruct</span> <span class="op">&gt;</span> out.log</span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co"># Move and gzip the database created </span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="fu">mv</span> RM<span class="pp">*</span> 01_RepMod</span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="bu">cd</span> 01_RepMod</span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="fu">tar</span> cvzf RepMod_rounds.tar.gz round-<span class="pp">*</span> LTR<span class="pp">*</span> tmp<span class="pp">*</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-r</span> round-<span class="pp">*</span> LTR<span class="pp">*</span> tmp<span class="pp">*</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="fu">gzip</span> families<span class="pp">*</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="fu">mv</span> <span class="va">$databasename</span><span class="pp">*</span> 01_RepMod/</span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="fu">cp</span>  ReferenceGenome.fasta <span class="va">$databasename</span>.fasta</span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a><span class="co"># Repeat Masker</span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a><span class="co">## -pa # of threads</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a><span class="co">## -a return alignment</span></span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a><span class="co">## -xsmall returns with masked lowercased </span></span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a><span class="co">## -gff creates the gene feature finding formatted output</span></span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a><span class="co">## -lib indicates the library, alternatively you could indicate the species</span></span>
<span id="cb2-32"><a href="#cb2-32" tabindex="-1"></a><span class="co">## -dir indicates output directory</span></span>
<span id="cb2-33"><a href="#cb2-33" tabindex="-1"></a><span class="co">## .fasta = reference genome</span></span>
<span id="cb2-34"><a href="#cb2-34" tabindex="-1"></a></span>
<span id="cb2-35"><a href="#cb2-35" tabindex="-1"></a><span class="ex">RepeatMasker</span> <span class="at">-pa</span> 24 <span class="at">-a</span> <span class="at">-xsmall</span> <span class="at">-gff</span> <span class="at">-lib</span> 01_RepMod/consensi.fa.classified <span class="at">-dir</span> 02_RepMask <span class="va">$databasename</span>.fasta</span>
<span id="cb2-36"><a href="#cb2-36" tabindex="-1"></a></span>
<span id="cb2-37"><a href="#cb2-37" tabindex="-1"></a><span class="co"># Prepare for use</span></span>
<span id="cb2-38"><a href="#cb2-38" tabindex="-1"></a><span class="fu">mkdir</span> 03_database</span>
<span id="cb2-39"><a href="#cb2-39" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1,4,5</span> 02_RepMask/<span class="va">$databasename</span>.fasta.out.gff <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-pi</span> <span class="at">-e</span> <span class="st">'s/^#.*\n//g'</span> <span class="op">&gt;</span> 03_database/ref2_<span class="va">$databasename</span>.gff.bed</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="remove-repeats-from-your-alignment">Remove repeats from your alignment<a class="anchor" aria-label="anchor" href="#remove-repeats-from-your-alignment"></a>
</h4>
<p>Based on the bed file created above, we then remove the repeats from
our alignment with samtools function <code>veiw</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Load modules</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="ex">module</span> load samtools/1.12 </span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co"># Make directories</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="fu">mkdir</span> repeats_removed</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">## Remove Repeats</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="ex">samtools</span> view sample_001.bam <span class="at">-b</span> <span class="at">-h</span> <span class="at">-o</span> /dev/null  <span class="at">-U</span> repeats_removed/sample_001.bam  <span class="at">-L</span> 03_database/ref2_<span class="va">$databasename</span>.gff.bed</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="filter-low-quality">Filter low quality<a class="anchor" aria-label="anchor" href="#filter-low-quality"></a>
</h3>
<p>To remove regions of low quality based on the MAPQ score. MAPQ is
equal to -10*log-base-10(Pr(mapping position is wrong)), rounded to the
nearest integer (<a href="https://samtools.github.io/hts-specs/SAMv1.pdf" class="external-link">see more
here</a>). For example, if we wanted to remove any site that had a 50%
chance of being mapped to the wrong position, we would set our filter to
4.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="op">-</span><span class="fl">10</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 4</span></span></code></pre></div>
<p>The calculation of the MAPQ score depends on the alignment software (
<a href="https://sequencing.qcfail.com/articles/mapq-values-are-really-useful-but-their-implementation-is-a-mess/" class="external-link">see
more here</a> ), therefore it is difficult to pinpoint the score needed
to remove reads that map to multiple locations.</p>
<p>Here we take a very stringent approach and remove any sites with that
have a 10% chance or more of being mapped to the wrong location and set
<code>-q</code> flag to 10.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Load modules</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="ex">module</span> load samtools/1.15 </span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co"># Make directories</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="fu">mkdir</span> filtered</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co">## Remove Repeats</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-b</span> <span class="at">-q</span> 10 repeats_removed/sample_001.bam <span class="op">&gt;</span> filtered/sample_001.bam</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="data-processing-with-nquack">Data processing with nQuack<a class="anchor" aria-label="anchor" href="#data-processing-with-nquack"></a>
</h2>
<div class="section level3">
<h3 id="step-1-prepare-">Step 1: Prepare.<a class="anchor" aria-label="anchor" href="#step-1-prepare-"></a>
</h3>
<p>Before running our model, your data must be converted to a
tab-seperated text file. To do this, we wrote a custom function
<code><a href="../reference/prepare_data.html">prepare_data()</a></code> which will convert your BAM file to a
tab-seperated text file with samtools. For this function, you must
supply the filename (without the .bam ending), the path to the directory
containing your BAM files, and the path to the directory where you want
your processed files to be saved.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Prepare many samples</span></span>
<span><span class="va">inpath</span> <span class="op">&lt;-</span> <span class="st">"filtered/"</span></span>
<span><span class="va">outpath</span> <span class="op">&lt;-</span> <span class="st">"Processed/"</span></span>
<span><span class="va">filelist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">inpath</span>, pattern <span class="op">=</span> <span class="st">"*.bam"</span> <span class="op">)</span></span>
<span><span class="va">filelist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".bam"</span>, <span class="st">""</span>, <span class="va">filelist</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span> <span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">filelist</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/prepare_data.html">prepare_data</a></span><span class="op">(</span><span class="va">filelist</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="va">inpath</span>, <span class="va">outpath</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level4">
<h4 id="warning-samtools-must-be-local">Warning: samtools must be local!<a class="anchor" aria-label="anchor" href="#warning-samtools-must-be-local"></a>
</h4>
<p>If you are working on your personal computer, just make sure samtools
is installed and callable as “samtools”. If you are working on a
cluster, you may need to install samtools locally. Though the location
of install may differ, here is how I installed samtools locally on UF’s
amazing <a href="https://www.rc.ufl.edu/about/hipergator/" class="external-link">HiPerGator</a> slurm
cluster:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">mkdir</span> bin</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="bu">cd</span> bin</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /apps/samtools/1.15/bin/samtools samtools</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="step-2-process">Step 2: Process<a class="anchor" aria-label="anchor" href="#step-2-process"></a>
</h3>
<p>Once your bam is converted into a tab-seperated text file, you must
process it and read it into the R environment. With this function, we
provide three options for filtering your data:</p>
<pre><code>(1) Total coverage filter 
(2) Allele coverage filter
(3) Allele frequency filter</code></pre>
<p>Total coverage can be filtered based on a minimum sequencing depth
and maximum sequencing depth quantile probability (coverage that falls
above the max.depth.quantile.prob will be removed). Allele coverage can
be filtered based on a sequencing error rate, where the coverage of each
allele must be more than the total coverage times the error rate:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>o</mi><mi>v</mi><mi>e</mi><mi>r</mi><mi>a</mi><mi>g</mi><mi>e</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>A</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>&gt;</mo><mi>C</mi><mi>o</mi><mi>v</mi><mi>e</mi><mi>r</mi><mi>a</mi><mi>g</mi><mi>e</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>A</mi><mo>+</mo><mi>B</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>*</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>e</mi><mi>r</mi><mi>r</mi><mi>o</mi><mi>r</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Coverage(A) &gt; Coverage(A + B)*(error)</annotation></semantics></math>,
but less than the total coverage times one minus the error rate:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>o</mi><mi>v</mi><mi>e</mi><mi>r</mi><mi>a</mi><mi>g</mi><mi>e</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>A</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>&lt;</mo><mi>C</mi><mi>o</mi><mi>v</mi><mi>e</mi><mi>r</mi><mi>a</mi><mi>g</mi><mi>e</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>A</mi><mo>+</mo><mi>B</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>*</mo><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>e</mi><mi>r</mi><mi>r</mi><mi>o</mi><mi>r</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Coverage(A) &lt; Coverage(A + B)*(1-error)</annotation></semantics></math>
. Finally, sites may be filtered based on the calculated allele
frequency, removing all sites below the lower bound,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>C</mi><mi>L</mi></msub><annotation encoding="application/x-tex">C_{L}</annotation></semantics></math>
, and above an a upper bound ,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>C</mi><mi>U</mi></msub><annotation encoding="application/x-tex">C_{U}</annotation></semantics></math>
. Allele frequency can be filtered based on a minimum and maximum allele
frequency.</p>
<p>Finally, to avoid data duplication, we randomly sample an allele with
equal probability at each site. The resulting data set includes the
total coverage per site and the coverage associated with a randomly
sampled allele.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Prepare many samples</span></span>
<span><span class="va">textfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"Processed/"</span>, pattern <span class="op">=</span> <span class="st">"*.txt"</span>, full.name <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">textfiles</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">temp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Processed/"</span>, <span class="va">textfiles</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="co"># File with full location</span></span>
<span>                       min.depth <span class="op">=</span> <span class="fl">2</span>, <span class="co"># Total coverage gilter</span></span>
<span>                       max.depth.quantile.prob <span class="op">=</span> <span class="fl">0.9</span>, <span class="co"># Total coverage filter</span></span>
<span>                       error <span class="op">=</span> <span class="fl">0.01</span>, <span class="co"># Allele Coverage Filter</span></span>
<span>                       trunc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="co"># Allele Frequency Filter</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/assign.html" class="external-link">assign</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".txt"</span>, <span class="st">""</span>, <span class="va">textfiles</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, <span class="va">temp</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level4">
<h4 id="when-should-you-filter-your-data">When should you filter your data?<a class="anchor" aria-label="anchor" href="#when-should-you-filter-your-data"></a>
</h4>
<div class="section level5">
<h5 id="coverage-filters-">Coverage filters.<a class="anchor" aria-label="anchor" href="#coverage-filters-"></a>
</h5>
<p>Increasing or decreasing the minimum and maximum coverage filter
parameters might be necessary for your data set. Total coverage can be
inspected based on the output from <code><a href="../reference/process_data.html">process_data()</a></code>, here we
call this output <code>xm</code>. If the x-axis on your coverage
histogram goes way above the targeted sequencing depth, you need to
decrease the maximum depth quantile probability.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">xm</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Error cutoffs</span></span>
<span><span class="co">### If I increase the sequence error rate, how many sites will likely be removed?</span></span>
<span><span class="va">new.e</span> <span class="op">&lt;-</span> <span class="fl">0.02</span> <span class="co"># 2 sites out of every 100 </span></span>
<span><span class="va">removes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">xm</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="va">xm</span><span class="op">[</span><span class="va">i</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">&lt;</span> <span class="op">(</span><span class="va">xm</span><span class="op">[</span><span class="va">i</span>,<span class="fl">1</span><span class="op">]</span><span class="op">*</span><span class="va">new.e</span><span class="op">)</span> <span class="op">|</span> <span class="va">xm</span><span class="op">[</span><span class="va">i</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">&gt;</span> <span class="op">(</span><span class="va">xm</span><span class="op">[</span><span class="va">i</span>,<span class="fl">1</span><span class="op">]</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">new.e</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">removes</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>  <span class="va">removes</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">removes</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="allele-frequency">Allele Frequency<a class="anchor" aria-label="anchor" href="#allele-frequency"></a>
</h5>
<p>Previous methods automatically truncated data by allele frequency,
removing any site with a frequencies below 0.1 or above 0.9. On first
pass, we suggest you do not truncate your data. However, once your data
is processed, you should plot your data to inspect.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert to allele frequency</span></span>
<span><span class="va">xi</span> <span class="op">&lt;-</span> <span class="va">xm</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">/</span><span class="va">xm</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">xi</span><span class="op">)</span></span></code></pre></div>
<p>For <i>Phytophthora infestans</i> (ENA:ERR1990235, Triploid), here is
the processed data with no truncation:</p>
<p><img src="../reference/figures/NoTrunc.png" width="125%" style="display: block; margin: auto;"></p>
<p>Notice the U-shaped ends? Well these should be removed prior to
ploidal estimation. We simply set <code>trunc = c(0.15,0.85)</code>
above and then reinspected our data:</p>
<p><img src="../reference/figures/Truncated.png" width="125%" style="display: block; margin: auto;"></p>
<p>One last note - if you expect you have 6x samples, be very careful
with your truncation, as the mean of one of the mixtures should be 0.16
for a hexaploid.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="comparing-nquack-to-nquire">Comparing nQuack to nQuire<a class="anchor" aria-label="anchor" href="#comparing-nquack-to-nquire"></a>
</h3>
<p>When you prepare data with nQuack, we use samtools mpileup, where
similarly to nQuire, we removes reads that did not align and PCR
duplicates (samtool flags: BAM_FUNMAP and BAM_FDUP). By default, nQuire
removes sites with coverage less than 10, as well as sites where
frequency is less than 0.2 or above 0.8. We allow users to modify these
parameters after the text file is creater rather than default to
arbitrary cutoffs. Here we mimic nQuire’s defaults for the triploid
mentioned above (<i>Phytophthora infestans</i>, ENA:ERR1990235). With
nQuack, you can replicate the nQuire data frame if desired.</p>
<p><img src="../reference/figures/DataCompare.png" width="125%" style="display: block; margin: auto;"></p>
<p>If you happen to like nQuire’s data preparation more than ours, uses
their data in our program. After processing samples with nQuire’s
<code>create</code> and <code>view</code> functions, the resulting txt
file can be read into R. To prepare the data frame for nQuack, reduce
the three column data frame to two columns by randomly sampling allele A
or B for every site. We created a function to help with this.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in nQuire txt file</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_nquire.html">process_nquire</a></span><span class="op">(</span><span class="st">"file.txt"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="what-about-denoising">What about ‘denoising’?<a class="anchor" aria-label="anchor" href="#what-about-denoising"></a>
</h3>
<p>Noise can get in the way of model selection, but identifying what
data is real and what is noise is difficult. Here we filter allele
frequencies with a normal + uniform mixture model. nQuack uses the
scaled probability of each data point belonging to each mixture model,
which is inferred in the expected maximization algorithm. We remove
allele frequencies where the probability of belonging to uniform mixture
is higher than their probability of belonging to any other mixture. We
also implement nQuire’s denoise method here, which utilizes the inferred
alpha parameter and a histogram of base frequencies to filter the
data.</p>
<p>This method is not without faults. Notably, this method does a poor
job of truncating the allele frequencies and this needs to be done prior
to attempting this method. For example, for the triploid mentioned above
(<i>Phytophthora infestans</i>, ENA:ERR1990235), if you only truncated
by <code>c(0.1, 0.9)</code> a peak toward the right is still
retained.</p>
<p><img src="../reference/figures/FancyDenoise88069-lowtrunc.png" width="125%" style="display: block; margin: auto;">
However, when you truncate the inital data set by
<code>c(0.15, 0.85)</code>, a much cleaner data set is returned by the
<code><a href="../reference/denoise_data.html">denoise_data()</a></code> function:</p>
<p><img src="../reference/figures/DenoiseFancy88069-highertrunc.png" width="125%" style="display: block; margin: auto;"></p>
<div class="section level4">
<h4 id="alternative-approach---bclean">Alternative approach - Bclean<a class="anchor" aria-label="anchor" href="#alternative-approach---bclean"></a>
</h4>
<p>When the shape and scale parameters associated with the beta
distribution are both less than 1, the distribution is U-shaped. We can
leverage this U-shaped distribution to remove noise. Here we utilize the
beta distribution with three mixtures representing cytotypes included in
nQuack and two mixtures representing a U-shaped distribution. We
constrained the first three mixtures to have shape and scale parameters
above 1, while the last two mixtures shape and scale are constrained to
be less than 1. With this implementation of expected maximization, we
utilizes the scaled probability of each data point belonging to each
mixture model to remove sites where the probability of belonging to a
U-shaped mixture is higher than the probability of belonging to any
other mixture. Due to the computational time needed to run the expected
maximization algorithm, by default, we simple calculate this probability
matrix with the E-step and do not run the complete algorithm.</p>
<p>This is a great alternative to allele truncation. Here is an example
of the <code><a href="../reference/Bclean.html">Bclean()</a></code> function applied to all nQuire’s samples
including <i>Phytophthora infestans</i> diploid (99189) and triploid
(88069), and <i>Saccharomyces cerevisiae</i> diploid (SRR3265396),
triploid (SRR3265389), and tetraploid (SRR3265401):</p>
<p><img src="../reference/figures/noTruncBclean.png" width="125%" style="display: block; margin: auto;"></p>
<p>This method can also be applied after truncation to catch any extra
peak:<br><img src="../reference/figures/TruncBClean.png" width="125%" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="bclean-and-denoise">Bclean and Denoise<a class="anchor" aria-label="anchor" href="#bclean-and-denoise"></a>
</h4>
<p>You can also add these two functions together, however this may
reduce your data set dramatically and impact on model inference has not
been thoroughly explored.</p>
<p><img src="../reference/figures/AllInThisTogether.png" width="125%" style="display: block; margin: auto;"></p>
<p>Now you are ready to run nQuack!</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by ML Gaynor.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
