<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Frequently Asked Questions • nQuack</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Frequently Asked Questions">
<meta name="description" content="Round-up of frequently asked questions.
">
<meta property="og:description" content="Round-up of frequently asked questions.
">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-YWYQ5CX5NC"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YWYQ5CX5NC');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nQuack</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/BasicExample.html">Basic Example</a></li>
    <li><a class="dropdown-item" href="../articles/DataPreparation.html">Data Preparation</a></li>
    <li><a class="dropdown-item" href="../articles/FasterExample.html">Faster Example</a></li>
    <li><a class="dropdown-item" href="../articles/ModelOptions.html">Model Options</a></li>
    <li><a class="dropdown-item" href="../articles/SimulateData.html">Simulate Data</a></li>
    <li><a class="dropdown-item" href="../articles/Outliers.html">Outliers</a></li>
    <li><a class="dropdown-item" href="../articles/FAQ.html">Frequently Asked Questions</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mgaynor1/nQuack/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Frequently Asked Questions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mgaynor1/nQuack/blob/HEAD/vignettes/FAQ.Rmd" class="external-link"><code>vignettes/FAQ.Rmd</code></a></small>
      <div class="d-none name"><code>FAQ.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="can-it-go-faster">Can it go faster?<a class="anchor" aria-label="anchor" href="#can-it-go-faster"></a>
</h2>
<p>See our <a href="https://mlgaynor.com/nQuack/articles/FasterExample.html" class="external-link">faster
example here</a>.</p>
</div>
<div class="section level2">
<h2 id="what-about-higher-ploidal-levels">What about higher ploidal levels?<a class="anchor" aria-label="anchor" href="#what-about-higher-ploidal-levels"></a>
</h2>
<p>Currently, nQuack is only coded to estimate the likelihood up to a
hexaploid (6x). However, our expectation-maximization code is able to
calculate the likelihood of any number of mixtures. Meaning, the code
can easily be modified to estimate if a ploidy level higher than 6x is
the most likely. Depending on your comfort in R, this may be simple or
overly confusing. I would start by reading through the <a href="https://mlgaynor.com/nQuack/articles/ModelOptions.html" class="external-link">Model
Option page</a> on the pkgdown site, let me know if you need additional
help with this.</p>
<p>I would caution using any approach based on site-based heterozygosity
to attempt to differentiation anything above 6 because the expectations
at this level are really difficult to differentiate. For example, if you
have a dodecaploid, you can have biallelic allele frequencies at 1/12,
2/12, 3/12, 4/12, 5/12, 6/12, 7/12, 8/12, 9/12, 10/12, 11/12 which would
be equivalent to the mean values of 0.083, 0.167, 0.25, 0.33, 0.417,
0.5, 0.583, 0.667, 0.75, 0.83, 0.913. Depending on your quality of data
and its coverage, these allele frequencies may be indistinguishable. I
would suggest <a href="https://mlgaynor.com/nQuack/articles/SimulateData.html" class="external-link">simulating
data</a> similar to what you have generated, here is a quick example of
how to simulate and view the expected allele frequencies for a
dodecaploid with our <a href="https://mlgaynor.com/nQuack/reference/sim.ind.BB.html" class="external-link">sim.ind.BB()
function</a>: <a href="https://gist.github.com/mgaynor1/7acfe13b526f94818469d535f3c30661" class="external-link uri">https://gist.github.com/mgaynor1/7acfe13b526f94818469d535f3c30661</a></p>
</div>
<div class="section level2">
<h2 id="can-i-identify-if-my-sample-is-an-allopolyploid-or-autopolyploid">Can I identify if my sample is an allopolyploid or
autopolyploid?<a class="anchor" aria-label="anchor" href="#can-i-identify-if-my-sample-is-an-allopolyploid-or-autopolyploid"></a>
</h2>
<p>Briefly, if you are using the model with alpha as a free parameter,
the alpha value will be indicative of the mode of inheritance. In
tetrasomic inheritance, you would expect that the model will assign
relatively equal proportions to the heterozygotes types, thus this would
indicate an autopolyploid. In an allopolyploid you would expect a higher
proportion at 0.5 due to disomic inheritance. You can check the
parameter output to see the calculated portions (or alpha values) by
running the expectation maximization outside of the wrapper functions.
Find more about about how to run the EM here: <a href="https://mlgaynor.com/nQuack/articles/ModelOptions.html" class="external-link uri">https://mlgaynor.com/nQuack/articles/ModelOptions.html</a></p>
<p>Here is an example of how to run the EM step: <a href="https://gist.github.com/mgaynor1/e5a4a9fb2eaba8aabbcf6ff1ef2075f1" class="external-link uri">https://gist.github.com/mgaynor1/e5a4a9fb2eaba8aabbcf6ff1ef2075f1</a></p>
<p>The best explanation of this approach can be found in Rowan’s
preprint here: Schley RS, Piñeiro R, Nicholls J, Gaynor ML, Lewis GP,
Pezzini FF, Dexter KG, Kider C, Pennington RT, and Twyford AD. The
frequency and importance of polyploidy in tropical trees. In Review, New
Phytologist. Preprint available, <a href="https://doi.org/10.1101/2025.07.28.667178" class="external-link">doi:
10.1101/2025.07.28.667178</a>.</p>
</div>
<div class="section level2">
<h2 id="what-about-rna-seq-hi-c-or-pacbio-hifi">What about RNA-seq? HI-C? or PacBio HIFI?<a class="anchor" aria-label="anchor" href="#what-about-rna-seq-hi-c-or-pacbio-hifi"></a>
</h2>
<p>We have not tested this method on RNA-seq. In theory, this could
potentially work on RNA-seq data, however, I am not sure how
allele-specific expression bias would impact the inference or if there
would be enough signal in the data to accurately predict ploidal level
for a sample. I suggest testing this approach with samples with
known-ploidy level and RNA-seq data to determine if this approach is
accurate for your data set - as indicated in our manuscript for any
application of this method.</p>
<p>I do not think think this method works for HI-C data.</p>
<p>This has never been tested on PacBio HIFI data. For high quality data
generated with PacBio, you may be able to infer mode of inheritance
based on a dotplot of syntenic anchor genes. For example, see Figure 3
(E and F) and the many many supplement plots from: <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC11785232/" class="external-link uri">https://pmc.ncbi.nlm.nih.gov/articles/PMC11785232/</a></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by ML Gaynor.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
