<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Model Options • nQuack</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Model Options">
<meta name="description" content="What models can you run with nQuack?
">
<meta property="og:description" content="What models can you run with nQuack?
">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-YWYQ5CX5NC"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YWYQ5CX5NC');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nQuack</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/BasicExample.html">Basic Example</a></li>
    <li><a class="dropdown-item" href="../articles/DataPreparation.html">Data Preparation</a></li>
    <li><a class="dropdown-item" href="../articles/FasterExample.html">Faster Example</a></li>
    <li><a class="dropdown-item" href="../articles/ModelOptions.html">Model Options</a></li>
    <li><a class="dropdown-item" href="../articles/SimulateData.html">Simulate Data</a></li>
    <li><a class="dropdown-item" href="../articles/Outliers.html">Outliers</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mgaynor1/nQuack/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Model Options</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mgaynor1/nQuack/blob/HEAD/vignettes/ModelOptions.Rmd" class="external-link"><code>vignettes/ModelOptions.Rmd</code></a></small>
      <div class="d-none name"><code>ModelOptions.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="site-based-heterozygosity">Site-based heterozygosity<a class="anchor" aria-label="anchor" href="#site-based-heterozygosity"></a>
</h2>
<p>At a biallelic site in a diploid with allele’s A and B, you expect to
sequence both alleles in equal proportions with 50% of sequences
representing A and 50% of sequences representing B. This simple idea is
the foundation of all <em>site-based heterozygosity</em> methods.</p>
<p>These approaches are based on biallelic single nucleotide
polymorphisms within an individual and the expected number of copies of
each allele.</p>
<p><img src="../reference/figures/TableofFreq.png" width="50%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="nquack">nQuack<a class="anchor" aria-label="anchor" href="#nquack"></a>
</h2>
<p>Like previous site-based heterozygosity approaches to predict ploidal
level, we use a mixture model approach to estimate the most likely
ploidal level.</p>
<p>nQuack provides multiple implementations of the expectation
maximization (EM) algorithm, including implementations with:</p>
<ol style="list-style-type: decimal">
<li>a normal distribution that mimics <a href="https://github.com/clwgg/nQuire" class="external-link">nQuire</a>,<br>
</li>
<li>a corrected normal distribution,<br>
</li>
<li>a beta distribution,<br>
</li>
<li>a beta-binomial distribution.</li>
</ol>
<p>Each of the implementations are provided with and without uniform
mixtures, for a total of eight implementations.</p>
<p>Here we details these available implementations and how to use
them.</p>
<div class="section level4">
<h4 id="model-types">Model types<a class="anchor" aria-label="anchor" href="#model-types"></a>
</h4>
<p>nQuire’s implementation of expectation maximization for a normal
distribution had two types of models: “free” where all variable were
estimated and “fixed” where only variance was estimated. nQuire then
calculate the log-likelihood ratio between the free and fixed models to
identify if a sample was a diploid, triploid, or tetraploid. The
simplest approach to expand this to higher ploidal levels is to just add
these distribution to the fixed mixtures examined.</p>
<p><img src="../reference/figures/Distributions.png" width="50%" style="display: block; margin: auto;"></p>
<p>However, mixture models have additional components that might be
helpful when predicting ploidal level. The figure below shows the basic
components of a mixture model which include mean
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>μ</mi><annotation encoding="application/x-tex">\mu</annotation></semantics></math>),
variance
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>σ</mi><annotation encoding="application/x-tex">\sigma</annotation></semantics></math>),
and proportion (or alpha,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>α</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math>).
Of these components, we only have expectations for the mean, which are
the expected frequencies of each allele, as seen in the table above. We
know that proportions may differ for an allopolyploid compared to an
autopolyploid, so we were interested in exploring models where alpha is
free.</p>
<p><img src="../reference/figures/Figure02-ModelDesign.png" width="100%" style="display: block; margin: auto;"></p>
<p>We ended up coding each implementation to provide estimates for all
parameters free (<code>type = 'free'</code>), only alpha free
(<code>type = 'fixed'</code>), only alpha and variance free
(<code>type = 'fixed_2'</code>), and only variance free
(<code>type ='fixed_3</code>). nQuire’s fixed models only predict
variance and resemble our <code>type = 'fixed_3'</code>
implementation.</p>
</div>
<div class="section level3">
<h3 id="additional-arguments-">Additional arguments.<a class="anchor" aria-label="anchor" href="#additional-arguments-"></a>
</h3>
<p>For each expectation maximization function, a starting parameter list
(<code>parm.list</code>) must be provided which includes
<code>avec</code>, <code>mvec</code>, and <code>svec</code>, which
represent
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>α</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>μ</mi><annotation encoding="application/x-tex">\mu</annotation></semantics></math>,
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>σ</mi><annotation encoding="application/x-tex">\sigma</annotation></semantics></math>
for each mixture. The length of each input <code>vec</code>tor should be
equal, unless the distribution includes a uniform mixture, then
<code>avec</code> should include a value for the uniform distribution.
All values must be greater than 0 and less than or equal to 1, with
<code>sum(avec)</code> equal to 1.</p>
<p>For the normal and beta distributions, allele frequency should be
provided (<code>xi</code>), however for the beta-binomial mixtures we
expect total coverage and coverage for a randomly sampled allele
(<code>xm</code>). Convergence of the algorithm occurs after the set
<code>niters</code> or iterates, or when the difference between the
augmented log-likelihood values is less than <code>epsilon</code>. We
also allow truncation of the probability density function to match any
allele frequency truncation that may have been done during data cleaning
(<code>trunc</code>).</p>
<p>Some notes:</p>
<ul>
<li>The equivalent to the nQuire model is our <code><a href="../reference/emstepNU.html">emstepNU()</a></code>
function.<br>
</li>
<li>Implementations with Nelder-Mead numerical optimization are unable
to estimate a single alpha value, therefore diploid mixtures cannot be
estimated from <code>emstepB(type =  "fixed")</code>and
<code>emstepBB(type = "fixed")</code> and the log-likelihood should be
calculated based on the sum of the probability density functions of
these mixtures.<br>
</li>
<li>Both the beta and beta-binomial use Nelder-Mead numerical
optimization, so they are computationally ‘slow’.<br>
</li>
<li>We provide packaged functions to apply these methods below.</li>
</ul>
<p>Expectation maximization implementations can be found here:</p>
<ul>
<li>Normal: <code><a href="../reference/emstepNA.html">emstepNA()</a></code>
</li>
<li>Normal with uniform: <code><a href="../reference/emstepNUA.html">emstepNUA()</a></code>
</li>
<li>nQuire’s normal: <code><a href="../reference/emstepN.html">emstepN()</a></code>
</li>
<li>nQuire’s normal with uniform: <code><a href="../reference/emstepNU.html">emstepNU()</a></code>
</li>
<li>Beta: <code><a href="../reference/emstepB.html">emstepB()</a></code>
</li>
<li>Beta with uniform: <code><a href="../reference/emstepBU.html">emstepBU()</a></code>
</li>
<li>Beta-binomial: <code><a href="../reference/emstepBB.html">emstepBB()</a></code>
</li>
<li>Beta-binomial with uniform: <code><a href="../reference/emstepBBU.html">emstepBBU()</a></code>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="mixture-model-implementation-">Mixture model implementation.<a class="anchor" aria-label="anchor" href="#mixture-model-implementation-"></a>
</h2>
<p>As we do not expect anyone to program the starting parameters needed
to run each mixture model, so we have designed a few functions to
help.</p>
<div class="section level3">
<h3 id="implementation">Implementation<a class="anchor" aria-label="anchor" href="#implementation"></a>
</h3>
<ul>
<li><code><a href="../reference/quackNormal.html">quackNormal()</a></code></li>
<li><code><a href="../reference/quackNormalNQ.html">quackNormalNQ()</a></code></li>
<li><code><a href="../reference/quackBeta.html">quackBeta()</a></code></li>
<li><code><a href="../reference/quackBetaBinom.html">quackBetaBinom()</a></code></li>
</ul>
<p>Each of these functions can run up to 32 mixture models each. A
matrix with total coverage and coverage for a randomly sampled allele
(<code>xm</code>) should be supplied to each of these functions. These
functions can run in parallel if resources are avaliable.</p>
<p>The input parameters for each mixture are based on the table above
with variance equal to 0.01 and alpha divided equally among the expected
frequencies. For example, for a triploid:
<code>mvec = c(0.33, 0.67)</code>, <code>avec=c(0.50, 0.50)</code>, and
<code>svec=c(0.01, 0.01)</code>. When a uniform mixture is included, we
allocate 0.1 of the avec to this mixture, therefore,
<code>avec = c(0.45, 0.45, 0.1)</code>. For each function, the input
mean and variance can be transformed based on sequencing error rate and
sequencing overdispersion parameter (see supplemental methods for more
information). Variance can also be set to 0.001.</p>
<ul>
<li><code><a href="../reference/bestquack.html">bestquack()</a></code></li>
</ul>
<p>Sometimes you may not want to run 32 mixture models, so we created a
function that lets you pick the distribution, type, and mixtures
included.</p>
</div>
<div class="section level3">
<h3 id="interpretation">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation"></a>
</h3>
<p>Based on the above functions, users are able to obtain the BIC
scores, log-likelihood, and log-likelihood ratios (or
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Δ</mi><mo>log</mo><mi>L</mi></mrow><annotation encoding="application/x-tex">\Delta\log{L}</annotation></semantics></math>)
of each mixture. For BIC and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>Δ</mi><mo>log</mo><mi>L</mi></mrow><annotation encoding="application/x-tex">\Delta\log{L}</annotation></semantics></math>,
the lowest value is the most likely model. Based on log-likelihood, the
highest value is the most likely model.</p>
<p>We recommend using the BIC which is penalized for both sample size
and number of parameters estimated. BIC will lead to less errors in
model selection (see Dennis et al. 2019).</p>
<p>To select the best model based on the model outputs, we provide the
function <code><a href="../reference/quackit.html">quackit()</a></code>.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by ML Gaynor.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
