<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Process data - Step 2 — process_data • nQuack</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Process data - Step 2 — process_data"><meta name="description" content="Based on the file generated with prepare_data(), which contains the total depth and sequencing coverage for each
nucleotide (A, C, G, and T), this function remove all but single nucelotide polymorphisms.
When supplied, this function will filter on coverage or  allele frequency. To filter by total coverage,
a user must supply the min.depth and max.depth.quantile.prob. If an error is provided,
sites will be retained where allele coverage is greater than the sequencing error rate times
the total coverage, but less than one minus the sequencing error rate times the total coverage.
Lastly, based on trunc, allele frequencies will be filtered based on a provided lower
and upper bound. Finally, the function samples a single allele frequency per site to avoid data duplication."><meta property="og:description" content="Based on the file generated with prepare_data(), which contains the total depth and sequencing coverage for each
nucleotide (A, C, G, and T), this function remove all but single nucelotide polymorphisms.
When supplied, this function will filter on coverage or  allele frequency. To filter by total coverage,
a user must supply the min.depth and max.depth.quantile.prob. If an error is provided,
sites will be retained where allele coverage is greater than the sequencing error rate times
the total coverage, but less than one minus the sequencing error rate times the total coverage.
Lastly, based on trunc, allele frequencies will be filtered based on a provided lower
and upper bound. Finally, the function samples a single allele frequency per site to avoid data duplication."><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-YWYQ5CX5NC"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YWYQ5CX5NC');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">nQuack</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/BasicExample.html">Basic Example</a></li>
    <li><a class="dropdown-item" href="../articles/DataPreparation.html">Data Preparation</a></li>
    <li><a class="dropdown-item" href="../articles/ModelOptions.html">Model Options</a></li>
    <li><a class="dropdown-item" href="../articles/SimulateData.html">Simulate Data</a></li>
    <li><a class="dropdown-item" href="../articles/Outliers.html">Outliers</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Process data - Step 2</h1>
      <small class="dont-index">Source: <a href="https://github.com/mgaynor1/nQuack/blob/HEAD/R/process_data.R" class="external-link"><code>R/process_data.R</code></a></small>
      <div class="d-none name"><code>process_data.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Based on the file generated with <code><a href="prepare_data.html">prepare_data()</a></code>, which contains the total depth and sequencing coverage for each
nucleotide (A, C, G, and T), this function remove all but single nucelotide polymorphisms.
When supplied, this function will filter on coverage or  allele frequency. To filter by total coverage,
a user must supply the <code>min.depth</code> and <code>max.depth.quantile.prob</code>. If an <code>error</code> is provided,
sites will be retained where allele coverage is greater than the sequencing error rate times
the total coverage, but less than one minus the sequencing error rate times the total coverage.
Lastly, based on <code>trunc</code>, allele frequencies will be filtered based on a provided lower
and upper bound. Finally, the function samples a single allele frequency per site to avoid data duplication.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">process_data</span><span class="op">(</span></span>
<span>  <span class="va">file</span>,</span>
<span>  min.depth <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  max.depth.quantile.prob <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  error <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  trunc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-file">file<a class="anchor" aria-label="anchor" href="#arg-file"></a></dt>
<dd><p>Output txt file created with <code><a href="prepare_data.html">prepare_data()</a></code>.</p></dd>


<dt id="arg-min-depth">min.depth<a class="anchor" aria-label="anchor" href="#arg-min-depth"></a></dt>
<dd><p>Minimum sequencing depth, default as 2.</p></dd>


<dt id="arg-max-depth-quantile-prob">max.depth.quantile.prob<a class="anchor" aria-label="anchor" href="#arg-max-depth-quantile-prob"></a></dt>
<dd><p>Maximum sequencing depth quantile cut off, default = 0.9.</p></dd>


<dt id="arg-error">error<a class="anchor" aria-label="anchor" href="#arg-error"></a></dt>
<dd><p>Sequencing error rate. If an <code>error</code> is provided,
sites will be retained where allele coverage is greater than the sequencing error rate times
the total coverage, but less than one minus the sequencing error rate times the total coverage.</p></dd>


<dt id="arg-trunc">trunc<a class="anchor" aria-label="anchor" href="#arg-trunc"></a></dt>
<dd><p>List of two values representing the lower and upper bounds, \(c_{L}\) and \(c_{U}\) which are used to filter allele frequencies.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Numeric matrix with total coverage and coverage for a randomly sampled allele.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by ML Gaynor.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

