# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @title E-Step for Expectation Maximization - Beta + Beta + Beta Distribution
#' @description This is used in the `Bclean()` function. Here we complete
#' the E-Step and calculate the log-likelihood. Modifications include a correction for
#' the truncated distribution.
#' @param parmlist A list containing initial alpha, mean, and variance.
#' @param xi List of observations, in this case allele frequencies.
#' @param trunc List of two values representing the lower and upper bounds, $c_{L}$ and $c_{U}$.
estepB3 <- function(parmlist, xi, trunc) {
    .Call(`_nQuack_estepB3`, parmlist, xi, trunc)
}

#' @title Expectation maximization - Beta + Beta + Beta Distribution
#'
#' @description This function is made for the `Bclean()` function and preforms expectation maximization with Nelder-Mead
#' numerical optimization for beta distribution.
#'
#' @param parmlist A list containing initial alpha, mean, and variance.
#' @param xi Matrix where the first column is total coverage and the second is the count of base A or B.
#' @param niter Max number of iterates.
#' @param epsilon Epsilon value for convergence tolerance. When the absolute delta log-likelihood is
#'    below this value, convergence is reached.
#' @param trunc List of two values representing the lower and upper bounds, $c_{L}$ and $c_{U}$.
#'
#' @returns List of elements including the negative log likelihood, the number of iterates,
#'  and the optimized parameter values.
#'
#'
emstepB3 <- function(parmlist, xi, niter, epsilon, trunc) {
    .Call(`_nQuack_emstepB3`, parmlist, xi, niter, epsilon, trunc)
}

#' @title Expectation maximization - Normal Distribution
#'
#' @description This function calculates the log-likelihood using
#'  the expectation maximization algorithm with the Normal Distribution.
#'  This code follows nQuire and does not use an augmented likelihood.
#'
#' @param parmlist A list containing initial alpha, mean, and variance values.
#' @param xi List of observations, in this case allele frequencies.
#' @param niter Max number of iterates.
#' @param epsilon Epsilon value for convergence tolerance. When the absolute delta log-likelihood is
#'    below this value, convergence is reached.
#' @param trunc List of two values representing the lower and upper bounds, $c_{L}$ and $c_{U}$.
#' @param type String indicating model type. Options: "free" (estimated parameter(s): alpha, mean, and variance), "fixed" (estimated parameter(s): alpha),
#' "fixed_2" (estimated parameter(s): alpha and variance), or "fixed_3" (estimated parameter(s): variance).
#'
#' @returns List of elements including the log-likelihood, the number of iterates,
#' and the optimized parameter values.
#'
emstepN <- function(parmlist, xi, niter, epsilon, trunc, type = "free") {
    .Call(`_nQuack_emstepN`, parmlist, xi, niter, epsilon, trunc, type)
}

#' @title Expectation maximization - Normal Distribution
#'
#' @description This function calculates the log-likelihood using
#'  the expectation maximization algorithm with the Normal Distribution.
#'  This code is not identical to nQuire and uses an augmented likelihood.
#'
#' @param parmlist A list containing initial alpha, mean, and variance values.
#' @param xi List of observations, in this case allele frequencies.
#' @param niter Max number of iterates.
#' @param epsilon Epsilon value for convergence tolerance. When the absolute delta log-likelihood is
#'    below this value, convergence is reached.
#' @param trunc List of two values representing the lower and upper bounds, $c_{L}$ and $c_{U}$.
#' @param type String indicating model type. Options: "free" (estimated parameter(s): alpha, mean, and variance), "fixed" (estimated parameter(s): alpha),
#' "fixed_2" (estimated parameter(s): alpha and variance), or "fixed_3" (estimated parameter(s): variance).
#'
#' @returns List of elements including the log-likelihood, the number of iterates,
#' and the optimized parameter values.
#'
emstepNA <- function(parmlist, xi, niter, epsilon, trunc, type = "free") {
    .Call(`_nQuack_emstepNA`, parmlist, xi, niter, epsilon, trunc, type)
}

#' @title Expectation maximization - Normal and Uniform Distribution
#'
#' @description This function calculates the log-likelihood using
#'  the expectation maximization algorithm with the Normal-Uniform Distribution.
#'  This code follows nQuire and does not use an augmented likelihood.
#'
#' @param parmlist A list containing initial alpha, mean, and variance values. The list of alpha must include a proportion for the uniform mixture.
#' @param xi List of observations, in this case allele frequencies.
#' @param niter Max number of iterates.
#' @param epsilon Epsilon value for convergence tolerance. When the absolute delta log-likelihood is
#'    below this value, convergence is reached.
#' @param trunc List of two values representing the lower and upper bounds, $c_{L}$ and $c_{U}$.
#' @param type String indicating model type. Options: "free" (estimated parameter(s): alpha, mean, and variance), "fixed" (estimated parameter(s): alpha),
#' "fixed_2" (estimated parameter(s): alpha and variance), or "fixed_3" (estimated parameter(s): variance).
#'
#' @returns List of elements including the log-likelihood, the number of iterates,
#' and the optimized parameter values.
#'
emstepNU <- function(parmlist, xi, niter, epsilon, trunc, type = "free") {
    .Call(`_nQuack_emstepNU`, parmlist, xi, niter, epsilon, trunc, type)
}

#' @title Expectation maximization - Normal Distribution
#'
#' @description This function calculates the log-likelihood using
#'  the expectation maximization algorithm with the Normal-Uniform Distribution.
#'  This code is not identical to nQuire and uses an augmented likelihood.
#'
#' @param parmlist A list containing initial alpha, mean, and variance values. The list of alpha must include a proportion for the uniform mixture.
#' @param xi List of observations, in this case allele frequencies.
#' @param niter Max number of iterates.
#' @param epsilon Epsilon value for convergence tolerance. When the absolute delta log-likelihood is
#'    below this value, convergence is reached.
#' @param trunc List of two values representing the lower and upper bounds, $c_{L}$ and $c_{U}$.
#' @param type String indicating model type. Options: "free" (estimated parameter(s): alpha, mean, and variance), "fixed" (estimated parameter(s): alpha),
#' "fixed_2" (estimated parameter(s): alpha and variance), or "fixed_3" (estimated parameter(s): variance).
#'
#' @returns List of elements including the log-likelihood, the number of iterates,
#' and the optimized parameter values.
#'
emstepNUA <- function(parmlist, xi, niter, epsilon, trunc, type = "free") {
    .Call(`_nQuack_emstepNUA`, parmlist, xi, niter, epsilon, trunc, type)
}

#' @title Expectation maximization - Beta Distribution
#'
#' @description This function calculates the log-likelihood using
#'  the expectation maximization algorithm with Nelder-Mead numerical optimization
#'  and a beta distribution.
#'
#' @param parmlist A list containing initial alpha, mean, and variance values.
#' @param xi List of observations, in this case allele frequencies.
#' @param niter Max number of iterates.
#' @param epsilon Epsilon value for convergence tolerance. When the absolute delta log-likelihood is
#'    below this value, convergence is reached.
#' @param trunc List of two values representing the lower and upper bounds, $c_{L}$ and $c_{U}$.
#' @param type String indicating model type. Options: "free" (estimated parameter(s): alpha, mean, and variance), "fixed" (estimated parameter(s): alpha),
#' "fixed-2" (estimated parameter(s): alpha and variance), or "fixed-3" (estimated parameter(s): variance).
#'  If avec is length of 1, fixed and fixed-3 will not be able to return a log-likelihood.
#'
#' @returns List of elements including the log likelihood, the negative log likelihood, the number of iterates,
#'  and the optimized parameter values.
#'
emstepB <- function(parmlist, xi, niter, epsilon, trunc, type = "free") {
    .Call(`_nQuack_emstepB`, parmlist, xi, niter, epsilon, trunc, type)
}

#' @title Expectation maximization - Beta-Binomial Distribution
#'
#' @description This function calculates the negative log-likelihood using
#'  the expectation maximization algorithm with Nelder-Mead numerical optimization
#'  and beta-binomial distribution.
#'
#' @param parmlist A list containing initial alpha, mean, and variance.
#' @param xm Matrix where the first column is total coverage and the second is the count of base A or B.
#' @param niter Max number of iterates.
#' @param epsilon Epsilon value for convergence tolerance. When the absolute delta log-likelihood is
#'    below this value, convergence is reached.
#' @param trunc List of two values representing the lower and upper bounds, $c_{L}$ and $c_{U}$.
#' @param type String indicating "Free" or "Fixed".
#'
#' @returns List of elements including the negative log likelihood, the number of iterates,
#'  and the optimized parameter values.
#'
#'
emstepBB <- function(parmlist, xm, niter, epsilon, trunc, type = "free") {
    .Call(`_nQuack_emstepBB`, parmlist, xm, niter, epsilon, trunc, type)
}

#' @title Expectation maximization - Beta-Binomial and Uniform Distributions
#'
#' @description This function calculates the log-likelihood using
#'  the expectation-maximization algorithm with Nelder-Mead numerical optimization
#'  and beta distribution with one uniform mixture.
#'
#' @param parmlist A list containing initial alpha, mean, and variance values.
#' @param xm Matrix where the first column is total coverage and the second is the count of base A or B.
#' @param niter Max number of iterates.
#' @param epsilon Epsilon value for convergence tolerance. When the absolute delta log-likelihood is
#'    below this value, convergence is reached.
#' @param trunc List of two values representing the lower and upper bounds, $c_{L}$ and $c_{U}$.
#' @param type String indicating model type. Options: "free" (estimated parameter(s): alpha, mean, and variance), "fixed" (estimated parameter(s): alpha),
#' "fixed-2" (estimated parameter(s): alpha and variance), or "fixed-3" (estimated parameter(s): variance).
#'  If avec is length of 1, fixed and fixed-3 will not be able to return a log-likelihood.
#'
#' @returns List of elements including the log likelihood, the negative log likelihood, the number of iterates,
#'  and the optimized parameter values.
#'
emstepBBU <- function(parmlist, xm, niter, epsilon, trunc, type = "free") {
    .Call(`_nQuack_emstepBBU`, parmlist, xm, niter, epsilon, trunc, type)
}

#' @title Expectation maximization - Beta and Uniform Distributions
#'
#' @description This function calculates the log-likelihood using
#'  the expectation maximization algorithm with Nelder-Mead numerical optimization
#'  and beta distribution with one uniform mixture.
#'
#' @param parmlist A list containing initial alpha, mean, and variance values.
#' @param xi List of observations, in this case allele frequencies.
#' @param niter Max number of iterates.
#' @param epsilon Epsilon value for convergence tolerance. When the absolute delta log-likelihood is
#'    below this value, convergence is reached.
#' @param trunc List of two values representing the lower and upper bounds, $c_{L}$ and $c_{U}$.
#' @param type String indicating model type. Options: "free" (estimated parameter(s): alpha, mean, and variance), "fixed" (estimated parameter(s): alpha),
#' "fixed_2" (estimated parameter(s): alpha and variance), or "fixed_3" (estimated parameter(s): variance).
#'  If avec is length of 1, fixed and fixed_3 will not be able to return a log-likelihood.
#'
#' @returns List of elements including the log likelihood, the negative log likelihood, the number of iterates,
#'  and the optimized parameter values.
#'
emstepBU <- function(parmlist, xi, niter, epsilon, trunc, type = "free") {
    .Call(`_nQuack_emstepBU`, parmlist, xi, niter, epsilon, trunc, type)
}

#' Calculate Alpha and Beta from Mean and Variance
#'
#' @param mu Mean.
#' @param var Variance.
#'
#' @return Numeric vector of alpha and beta.
alphabetacalc <- function(mu, var) {
    .Call(`_nQuack_alphabetacalc`, mu, var)
}

#' Vector-based - Calculate Alpha and Beta from Mean and Variance
#'
#' @param mu Vector of mean.
#' @param var Vector of variance.
#'
#' @return Numeric matrix of alpha and beta.
alphabetacalcvec <- function(mu, var) {
    .Call(`_nQuack_alphabetacalcvec`, mu, var)
}

#' Calculate Alpha and Beta from Mean, Tau, and Error rate.
#'
#' @param mu Mean.
#' @param tau Overdispersion parameter. Ranges from 0 to 1, where 0 indicates less overdispersion and 1 indicates high overdispersion.  Here tau must be greater than 0.
#' @param error Sequencing error rate.
#'
#' @return Numeric vector of alpha and beta.
alphabetacalctau <- function(mu, tau, error) {
    .Call(`_nQuack_alphabetacalctau`, mu, tau, error)
}

#' Vector-based - Calculate Alpha and Beta from Mean, Tau, and Error rate.
#'
#' @param mu Vector of mean.
#' @param tau Overdispersion parameter. Ranges from 0 to 1, where 0 indicates less overdispersion and 1 indicates high overdispersion.  Here tau must be greater than 0.
#' @param error Sequencing error rate. Ranges from 0 to 1.
#'
#' @return Numeric matrix of alpha and beta.
alphabetacalctauvec <- function(mu, tau, error) {
    .Call(`_nQuack_alphabetacalctauvec`, mu, tau, error)
}

#' @title Prepare data - Step 1
#'
#' @description This function transforms a BAM file into a text file.
#'   Specifically, this function uses [samtools mpileup](http://www.htslib.org/doc/samtools-mpileup.html)
#'   to translate your BAM into a tab-separated file. We then filter this file to remove indels and deletions.
#'   When running this function, a temporary folder will be created (named 'temp/'), however this folder will be
#'   removed once the process is complete.
#'
#' @details Warning, due to the processing time needed for samtools mpileup,
#'   this step may take some time. This function also requires samtools to be located locally. Please see
#'   our [Data Preparation](mlgaynor.com/nQuack/DataPreparation.html) article for more information. Warning, this writes a temporary folder
#'   titled 'temp'. If you want to run multiple samples at once, we suggest you set the working directory to separate locations to ensure that
#'   your temp folder/files are not overwritten.
#'
#' @param name File name without the suffix. For example, if your file is called "frog.bam", this input should be "frog".
#' @param inpath Location of input file.
#' @param outpath Location for output file.
#'
#' @returns Writes text file with the following columns: chromosome, position, depth, A, C, G, and T.
#'
prepare_data <- function(name, inpath, outpath) {
    invisible(.Call(`_nQuack_prepare_data`, name, inpath, outpath))
}

#' Calculate Alpha and Beta from Mean and Variance
#'
#' @param xm Matrix with total coverage and coverage at a randomly sampled allele.
#' @param n Length of matrix.
#'
#' @return Randomly sampled matrix.
resample_xm <- function(xm, n) {
    .Call(`_nQuack_resample_xm`, xm, n)
}

#' @title Data Preparation - Use nQuire's Data
#'
#' @description This function reduce a three column data frame to
#' two columns by randomly sampling allele A or B for every site. This is used in our function `process_nquire()`
#'
#' @param xm A matrix with three columns: Total Coverage, Counts for Allele A, and Counts for Allele B.
#'
#' @returns Numeric Matrix with total coverage and coverage for a randomly sampled allele.
nQuire_reformat <- function(xm) {
    .Call(`_nQuack_nQuire_reformat`, xm)
}

#' @name process_rcpp
#' @title Data Preparation - Matrix Filtering
#' @description Based on supplied matrix with total depth and sequencing coverage for each
#'  nucleotide (A, C, G, and T) this function remove all but single nucelotide polymorphisms.
#'  When supplied, this function will filter on coverage or  allele frequency. Finally, the function
#'  samples a single allele frequency per site to avoid data duplication.
#'
#' @param x Matrix with five columns: Depth, A, C, G, and T.
#' @param mindepth Minimum depth, default = 15.
#' @param maxprob Maximum depth quantile cut off, default = 0.9.
#' @param trunc List of two values representing the lower and upper bounds, $c_{L}$ and $c_{U}$.
#' @param error Sequencing error rate.
#'
#' @return Numeric Matrix with total coverage and coverage for a randomly sampled allele.
#'
process_rcpp <- function(x, mindepth, maxprob, trunc, error) {
    .Call(`_nQuack_process_rcpp`, x, mindepth, maxprob, trunc, error)
}

# Register entry points for exported C++ functions
methods::setLoadAction(function(ns) {
    .Call(`_nQuack_RcppExport_registerCCallable`)
})
